### PR TITLE
[video] Change Versions Art Handling

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

Make the parent movie and versions behave more like sets and movies with respect to art:

- when browsing movies in the library, the art scraped/set by the user for the parent movie is used.
- choosing art from library manage context menu or info dialog sets the parent movie art.
- when browsing the versions of a movie (choose dialog from library, versions as a subfolder of the movie, versions top-level navigation), the art scraped/set by user for the version is used. If some types of art is missing at the version level, the art of the parent movie is used instead. Parent movie art will usually exist, since in v21 the movie is still the core element of the library.
- browsing versions/choosing art from Versions manager shows the art of the versions only, with no fallback on the art of the parent movie, for clarity.

Overall this allows different art for the group and each version, but if version art is not provided, group art will be used for everything.

Technically this is done by checking the path/options of the item at key points, and adding a special property to disable the art fallback that usually takes place, except when editing art.

This PR doesn't address the current issues of the Manage context menu and the Info dialog when launched for a version of a movie (from versions as subfolder of a movie or versions top-level navigation). The problems go beyond art. In that context, the dialog should probably replace some functions of Versions Manager and versions/extas buttons should be hidden. It may be v22 work though, an easier fix would be to remove access to the dialog.

Two small related changes included:
- [fix] ignore art with url field empty: setting art to "no art" doesn't delete the record in the art table but sets the url to blank. Those records have to be ignored for fallback between version and parent movie to work properly.
- [cosmetic] move some code of GetMoviesByWhere() into GetFilter() where it belongs

## Motivation and context

Address some surprising behavior of art in master that are consequences of always using version art first, even when browsing movies in the library.
Attempt to provide a no-change solution for the most common case of single version for movies and small code change when there are multiple versions.

From the user point of view should feel similar to what will be available later in v22, although the implementation is not clean due to constraints of the current design.

This superseeds my previous attempts.

## How has this been tested?

Movies with single and multiple versions.
Changed art with context menu > choose art, with info dialog Choose art
Changed art with Versions Manage dialog.
Browsed the library by titles, from versions navigation, in the versions subfolders.

## What is the effect on users?

Intuitive behavior of art.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp [hunk 1]
```diff
@@ -847,7 +847,7 @@ void AddCurrentArtTypes(std::vector<std::string>& artTypes, const CVideoInfoTag&
   std::map<std::string, std::string> currentArt;
 
   if (tag.GetAssetInfo().GetId() >= 0)
-    db.GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, currentArt);
+    db.GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, false, currentArt);
```
Comment (ksooo) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp:850]: You know I do not like bool parameters in public interfaces. ;-) What does this new `false` tell me? I need to look up the method declaration in the respective header file to get to know. If you would use an enum class with speaking values, understanding this code would be much easier. Any change ti change the bool param to an enum class param?

File: xbmc/video/VideoThumbLoader.cpp [hunk 2]
```diff
@@ -397,7 +397,8 @@ bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)
   {
     m_videoDatabase->Open();
 
-    if (tag.GetAssetInfo().GetId() >= 0)
+    const CURL itemUrl{item.GetPath()};
+    if (itemUrl.HasOption("videoversionid"))
```
Comment (ksooo) [inline xbmc/video/VideoThumbLoader.cpp:401]: Why this change? Just want to understand...

File: xbmc/video/VideoThumbLoader.cpp [hunk 3]
```diff
@@ -400,7 +400,12 @@ bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)
     const CURL itemUrl{item.GetPath()};
     if (itemUrl.HasOption("videoversionid"))
     {
-      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, artwork))
+      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type,
+                                          item.HasProperty("noartfallbacktoowner") &&
+                                                  item.GetProperty("noartfallbacktoowner") == true
+                                              ? false
+                                              : true,
```
Comment (ksooo) [inline xbmc/video/VideoThumbLoader.cpp:407]: Not 100% sure, but imo the whole term could be shortened to `!item.GetProperty("noartfallbacktoowner").asBoolen(false)`.

Comment (CrystalP) [reply xbmc/video/VideoThumbLoader.cpp:407]: yes. that's convenient!

File: xbmc/video/VideoThumbLoader.cpp (hunk 2 above)
Comment (CrystalP) [reply xbmc/video/VideoThumbLoader.cpp:401]: This is how the code knows that we're browsing movies in the library and considering the movies as groups of versions (use "movie" art type) or using one of the paths that displays the individual versions (use "videoversion" art type): versions in folder, choose version dialog, versions navigation, versions manager dialog.

Same could be done with a GetPath() and looking for a videodb://movies/videoversions/ prefix.

It's not forward looking but works today. In v22 this will have to be redone completely, stuff moved around, ...

Comment (ksooo) [reply xbmc/video/VideoThumbLoader.cpp:401]: Okay, thanks for the explanation.

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp (hunk 1 above)
Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoInfo.cpp:850]: I saw this coming but am still thinking about a better way. Likely will be a different overload of GetArtForAsset with less parameters when fallback is not wanted. Inside videodb it would still be the same.

Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoInfo.cpp:850]: and some doxy...

Comment (ksooo) [reply xbmc/video/dialogs/GUIDialogVideoInfo.cpp:850]: > Likely will be a different overload of GetArtForAsset with less parameters

That would be even more elegant, yes.

File: xbmc/video/VideoThumbLoader.cpp (hunk 2 above)
Comment (CrystalP) [reply xbmc/video/VideoThumbLoader.cpp]: refactored into a method of CFileItem as the same condition will have to be used to fix stuff in info dialog / context menu opened on versions, if the concept of this PR is accepted.

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp (hunk 1 above)
Comment (CrystalP) [reply xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: the overload turned out cumbersome, then I simplified the function, in the end it's an enum.

File: xbmc/FileItem.cpp [hunk 4]
```diff
@@ -4273,3 +4273,14 @@ bool CFileItem::HasVideoExtras() const
   }
   return false;
 }
+
+bool CFileItem::IsVideoAssetNav() const
+{
+  // @todo maybe in the future look for prefix videodb://movies/videoversions in path instead
+  // @todo better encoding of video assets as path, they won't always be tied with movies.
+  const CURL itemUrl{GetPath()};
```
Comment (ksooo) [inline xbmc/FileItem.cpp:4281]: Maybe add a check for being a videodb item first? `if (IsVideoDb()) ...` This is cheap and will avoid parsing of the URL etc. in many cases.

File: xbmc/video/VideoDatabase.h [hunk 5]
```diff
@@ -401,6 +401,12 @@ const struct SDbTableOffsets DbMusicVideoOffsets[] =
   { VIDEODB_TYPE_INT, my_offsetof(CVideoInfoTag,m_iIdUniqueID)}
 };
 
+typedef enum
```
Comment (ksooo) [inline xbmc/video/VideoDatabase.h:404]: I would suggest an `enum class`. In general we want to get rid of classic enum's and typedef's if possible, because enum classes have some benefits, for example you can forward declare them.

File: xbmc/video/VideoThumbLoader.cpp [hunk 6]
```diff
@@ -397,9 +397,14 @@ bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)
...
 
-    if (tag.GetAssetInfo().GetId() >= 0)
+    // @todo unify asset path for other items path
+    if (item.IsVideoAssetNav())
     {
-      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId, tag.m_iDbId, tag.m_type, artwork))
+      if (m_videoDatabase->GetArtForAsset(tag.m_iFileId,
+                                          item.GetProperty("noartfallbacktoowner").asBoolean(false)
+                                              ? ARTFALLBACK_NONE
+                                              : ARTFALLBACK_PARENT,
```
Comment (ksooo) [inline xbmc/video/VideoThumbLoader.cpp:406]: I like. 👍

**Comment by CrystalP** [comment]
Thanks for the suggestions, applied!

**Comment by CrystalP** [comment]
Jenkins build this please

**Review by ksooo — APPROVED** [review]
LGTM
